packages/playground/data-liberation/src/import/WP_Topological_Sorter.php

```diff
@@ -0,0 +1,461 @@
+				);
+
+				CREATE UNIQUE INDEX IF NOT EXISTS idx_element_id ON %i (element_id);
+				CREATE INDEX IF NOT EXISTS idx_session_id ON %i (session_id);
+				CREATE INDEX IF NOT EXISTS idx_parent_id ON %i (parent_id);
+				CREATE INDEX IF NOT EXISTS idx_byte_offset ON %i (byte_offset);',
+				$table_name,
+				self::ELEMENT_TYPE_POST,
+				$table_name,
+				$table_name,
+				$table_name,
+				$table_name
+			);
+		} else {
+			// See wp_get_db_schema
+			$max_index_length = 191;
+
+			// MySQL, MariaDB.
+			$sql = $wpdb->prepare(
+				'CREATE TABLE IF NOT EXISTS %i (
+					id bigint(20) unsigned NOT NULL AUTO_INCREMENT,
+					session_id bigint(20) unsigned NOT NULL,
+					element_type tinyint(1) NOT NULL default %d,
+					element_id text NOT NULL,
+					parent_id text DEFAULT NULL,
+					byte_offset bigint(20) unsigned NOT NULL,
+					sort_order int DEFAULT 1,
+					PRIMARY KEY  (id),
+					KEY session_id (session_id),
+					KEY element_id (element_id(%d)),
+					KEY parent_id (parent_id(%d)),
+					KEY byte_offset (byte_offset)
+				) ' . $wpdb->get_charset_collate(),
+				self::get_table_name(),
+				self::ELEMENT_TYPE_POST,
+				$max_index_length,
+				$max_index_length
+			);
+		}
+
+		require_once ABSPATH . 'wp-admin/includes/upgrade.php';
+		dbDelta( $sql );
+
+		update_option( self::OPTION_NAME, self::DB_VERSION );
+	}
+
+	public static function is_sqlite() {
+		return defined( 'DB_ENGINE' ) && 'sqlite' === DB_ENGINE;
+	}
+
+	/**
+	 * Run in the 'plugins_loaded' action.
+	 */
+	public static function load() {
+		if ( self::DB_VERSION !== (int) get_site_option( self::OPTION_NAME ) ) {
+			// Used to update the database with dbDelta, if needed in the future.
+			self::activate();
+		}
+	}
+
+	/**
+	 * Run by register_deactivation_hook.
+	 */
+	public static function deactivate() {
+		global $wpdb;
+		$table_name = self::get_table_name();
+
+		// Drop the table.
+		$wpdb->query( $wpdb->prepare( 'DROP TABLE IF EXISTS %s', $table_name ) );
+
+		// Delete the option.
+		delete_option( self::OPTION_NAME );
+	}
+
+	/**
+	 * Run by register_uninstall_hook.
+	 */
+	public function reset() {
+		$this->orphan_post_counter = 0;
+		$this->last_post_id        = 0;
+		$this->sorted              = false;
+		$this->current_session     = null;
+		$this->total_categories    = 0;
+		$this->total_posts         = 0;
+		$this->current_item        = 0;
+	}
+
+	/**
+	 * Delete all rows for a given session ID.
+	 *
+	 * @param int $session_id The session ID to delete rows for.
+	 * @return int|false The number of rows deleted, or false on error.
+	 */
+	public function delete_session( $session_id ) {
+		global $wpdb;
+
+		return $wpdb->delete(
+			self::get_table_name(),
+			array( 'session_id' => $session_id ),
+			array( '%d' )
+		);
+	}
+
+	/**
+	 * Map a category to the index.
+	 *
+	 * @param int $byte_offset The byte offset of the category.
+	 * @param array $data The category data.
+	 */
+	public function map_category( $byte_offset, $data ) {
```

Comment:
map makes me think of map/reduce and a mapper callback. How about `insert_category` or `add_category` or `index_category`?